doc/developer/design/20250307_column_names.md

```diff
@@ -0,0 +1,178 @@
+
+```
+Explained Query:
+  Project (#2, #1{y}) // { arity: 2 }
+    Map ((#0{x} + 1)) // { arity: 3 }
+      ReadStorage materialize.public.t // { arity: 2 }
+
+Source materialize.public.t
+
+Target cluster: quickstart
+```
+
+## Success Criteria
+
+For an MIR expression `q`, define:
+
+```
+NAMESCORE(q) = |# of named column references|
+               ------------------------------
+                  |# of column references|
+```
+
+i.e., `NAMESCORE(q)` is the percentage of column references which we
+can attach a name to.
+
+We should expect to see a serious improvement in `NAMESCORE` across
+both our tests and customer queries.
+
+## Out of Scope
+
+We should _not_ expect `NAMESCORE(q) = 1.0` for all queries `q`.
+
+In particular, window functions may significantly complicate our
+ability to keep track of names.
+
+We should not reimplement all of the `Analysis` framework for LIR.
+
+## Solution Proposal
+
+We will need to keep track of more column name information. We propose
+doing so in a few positions:
+
+1. Keep track of column names all `ScalarExpr` constructors.
+
+2. Keep track of column names on `usize` column references in
+   `*RelationExpr`.
+
+In both cases, we will work with `Option<Rc<str>>`, being careful
+to generate a single reference for each string, i.e., we will do
+lightweight, reference-counted string internment to avoid tons of
+extra allocations.
+
+### Column names on `{Hir,Mir}ScalarExpr::Column`
+
+Extend `HirScalarExpr` and `MirScalarExpr` to optionally keep track of
+column names, i.e.:
+
+```rust
+pub enum HirScalarExpr {
+  Column(ColumnRef, Option<Rc<str>>),
+  CallUnary {
+    func: UnaryFunc,
+    expr: Box<HirScalarExpr>,
+    name: Option<Rc<str>>>,
+  }
+  // ...
+}
+
+impl HirScalarExpr {
+  pub fn name(&self) -> Option<Rc<str>> {
+    match self {
+      HirScalarExpr::Column(_, name)
+      | HirScalrExpr::CallUnary { name, .. }
+      | ... => name.clone()
+    }
+  }
+}
+```
+
+Keeping track of column names on _every_ expression makes it so we can
+remember names for SQL CTEs. Lowering would change to generate scalar
+expressions with names for:
+
+  - column references
+  - `SELECT expr AS name, ...`
+  - `CREATE VIEW v(name, ...) AS ...`
+  - `WITH t(name, ...), ...`
+  - `WITH MUTUALLY RECURSIVE t(name type, ...), ...)`
+
+### Column names in `{Hir,Mir}RelationExpr::{Project,Reduce,TopK}`
+
+Extend `HirScalarExpr` and `MirScalarExpr` to optionally keep track of
+the column name.
+
+```rust
+pub enum HirRelationExpr {
+    Project {
+        input: Box<HirRelationExpr>,
+        outputs: Vec<(usize, Option<Rc<str>>)>,
+    },
+    // ...
+}
+
+pub enum MirRelationExpr {
+    Project {
+        input: Box<MirRelationExpr>,
+        outputs: Vec<(usize, Option<Rc<str>>)>,
+    },
+    // ...
+}
+```
+
+The `Project` is the most critical, as projections appear at the top
+of nearly every `SELECT` query---the missing `succ_x` in the example
+above appears in a `Project`, for example.
```

Comment:
These Projects often go through `MapFilterProject` (e.g., `CanonicalizeMfp` eats up all the Maps, Filters, Projects, creates a `MapFilterProject`, then optimizes it, and and then converts it back to `Map`,`Filter`,`Project`), so we'll have to augment that also with column names.